src/NuGet.CommandLine/MultipartWebRequest.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Net;
+using System.Text;
+
+namespace NuGet.CommandLine
+{
+    /// <remarks>
+    /// Based on the blog post by Travis Illig at http://www.paraesthesia.com/archive/2009/12/16/posting-multipartform-data-using-.net-webrequest.aspx
+    /// </remarks>
+    public class MultipartWebRequest
+    {
+        private const string FormDataTemplate = "--{0}\r\nContent-Disposition: form-data; name=\"{1}\"\r\n\r\n{2}\r\n";
+        private const string FileTemplate = "--{0}\r\nContent-Disposition: form-data; name=\"{1}\"; filename=\"{2}\"\r\nContent-Type: {3}\r\n\r\n";
+        private readonly Dictionary<string, string> _formData;
+
+        private readonly List<PostFileData> _files;
+
+        public MultipartWebRequest()
+            : this(new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase))
+        {
+        }
+
+        public MultipartWebRequest(Dictionary<string, string> formData)
+        {
+            _formData = formData;
+            _files = new List<PostFileData>();
+        }
+
+        public void AddFormData(string key, string value)
+        {
+            _formData.Add(key, value);
+        }
+
+        public void AddFile(Func<Stream> fileFactory, string fieldName, long length, string contentType = "application/octet-stream")
+        {
+            _files.Add(new PostFileData
+            {
+                FileFactory = fileFactory,
+                FieldName = fieldName,
+                ContentType = contentType,
+                ContentLength = length
+            });
+        }
+
+        public void CreateMultipartRequest(WebRequest request)
+        {
+            string boundary = "---------------------------" + DateTime.Now.Ticks.ToString("x", CultureInfo.InvariantCulture);
```

Comment:
Use var here and in other places in the CR as applicable. Another example is using (Stream stream below)